### PR TITLE
Add Authentication [WIP]

### LINKED ISSUE #100: Implement open ID connect, relying party authentication
All requests that modify data, should be authenticated as a OIDC RP.

We should have a transitional env variable that disables this, so we can test it without authenticating.

### PR DESCRIPTION
**Pull request by coffeejunk**
Closes #100 

A [separate application](https://github.com/coffeejunk/crce-auth_server) allows users to register using Google OIDC.

Warden is used to authenticate a user before any destructive call on the `/resources` endpoints. The "authentication" is currently done with the UID and will subsequently have to be replaced with a real token.

TODOs:
- [ ] tests for authentication
- [ ] updated existing tests 
- [ ] add a "development" mode/flag that removes the need for authentication
- [ ] use token instead of user uid to authenticate a request
- [ ] (Google Group) Approval process
- [ ] Documentation
- [ ] Register in envelope the API consumer ID

Currently both the authentication application as well as CR are sharing one database. This is typically a bad idea for several reasons (esp. potential inconsistencies) however in this case it seemed opportune as only the `ApiConsumer` is really shared. Depending on how complex the registration and especially authorization part of the application become it might be better to have the applications use independent databases and let them communicate over the wire.

### CODE REVIEW

**Comment by leolara** [comment]
@coffeejunk 

Ok to the sign up application independent.

Ok to WIP, but make this a incremental PR part of a bigger problem. So the main function is just to authenticate, and save in the created/updated/deleted envelope the apiconsumer that did it.

All the Pub/Priv Key managament, publisher stuff and auto-signed keys should be done incrementally in other PRs once this one is merged.

So I think in your TODO list I would remove:
+ Publisher
+ Relation between Publisher and Envelopes

Also, what does it mean "real tokens"?

**Comment by coffeejunk** [comment]
By "real tokens" I meant, using a token for the authentication instead of the user's `uid` as is done right now.

**Comment by leolara** [comment]
@coffeejunk I added as a subtask to register in the envelope the API consumer ID. If you think it will make this PR too long, feel free to do it in another task

File: db/migrate/20170725143941_create_api_consumers.rb [hunk 1]
```diff
@@ -0,0 +1,12 @@
+class CreateApiConsumers < ActiveRecord::Migration
+  def change
+    create_table :api_consumers do |t|
+      t.string :name
+      t.string :email, null: false, index: true
+      t.string :provider, null: false
+      t.string :uid, null: false, index: true
```
Comment (andersoncardoso) [inline db/migrate/20170725143941_create_api_consumers.rb]: not sure from where the `uid` will come, but I'm guessing these should be unique, if that's the case maybe we can add a constraint to it.

Comment (leolara) [inline db/migrate/20170725143941_create_api_consumers.rb]: @andersoncardoso @coffeejunk I think the uid shoud be a autoincrement or something like that. we want to allow to extensibility, so currently implemeting OpenIDC but we would like to allow the user to authenticate in other form, so uid should be agnostic to the way the user authenticates.

Comment (coffeejunk) [inline db/migrate/20170725143941_create_api_consumers.rb]: the `uid` is (currently) the `uid` that we receive from google

Comment (leolara) [inline db/migrate/20170725143941_create_api_consumers.rb]: @coffeejunk if you look at the design 

https://docs.google.com/document/d/1LcZD8dJXeDnOrYtdrEZ6Dq3820mi2P0FUZNJL6Jd-Dg/edit#heading=h.9eb1b9xngsnr

the OpenIDC information is attribute (or a set of attributes) to the API customer. So if there is a OpenIDC id field you need to save that should be in another OpenIDC specfic field, that way we could add other means of authentication in the future

Also, Google is an OpenIDC provider but we would like it to work with other OpenIDC providers if possible, that does not mean that we implement the sign up for other providers now, but at least make it enough generic

**Comment by rsaksida** [comment]
Apparently this has been addressed with "publish on behalf":

**Pull request closed without merging**